Update Amazon SQS connector doc (#708)
wso2/docs-mi: en/docs/reference/connectors/amazonsqs-connector/1.x/amazonsqs-connector-1.x-example.md

# AmazonSQS Connector Example

The WSO2 Amazon SQS connector allows you to access the exposed Amazon SQS API from an integration sequence.

## What you'll build

This example explains how to use Amazon SQS Connector to create a queue in the Amazon SQS, send a message to the queue, forward it to Simple Stock Quote Service Backend, and send the response to the user. 

It has a single HTTP API resource, which is `sendToQueue`. 

<img src="{{base_path}}/assets/img/integrate/connectors/amazonsqs-connector.jpg" title="AmazonSQS-Connector" width="800" alt="AmazonSQS-Connector"/>

If you do not want to configure this yourself, you can simply [get the project](#get-the-project) and run it.

## Set up the environment 

> 1. Please follow the steps mentioned in the [Setting up the Amazon S3 Environment]({{base_path}}/reference/connectors/amazonsqs-connector/amazonsqs-connector-config) document to create an Amazon account and obtain access key ID and secret access key. Keep them saved to be used in the next steps.  

Reviewer: ```suggestion
1. Follow the steps mentioned in the [Setting up the Amazon S3 Environment]({{base_path}}/reference/connectors/amazonsqs-connector/amazonsqs-connector-config) document to create an Amazon account and obtain access key ID and secret access key. Keep them saved to be used in the next steps.  
```

> 1. Please follow the steps mentioned in the [Setting up the Amazon S3 Environment ]({{base_path}}/reference/connectors/amazonsqs-connector/amazonsqs-connector-config) document in order to create an Amazon account and obtain access key id and secret access key. Keep them saved to be used in the next steps.  

Reviewer: ```suggestion
1. Please follow the steps mentioned in the [Setting up the Amazon S3 Environment]({{base_path}}/reference/connectors/amazonsqs-connector/amazonsqs-connector-config) document to create an Amazon account and obtain access key ID and secret access key. Keep them saved to be used in the next steps.  
```

> 2. In this example we will be using XPath 2.0 which needs to be enabled in the product as shown below before starting the integration service. 

Reviewer: ```suggestion
2. In this example, we will be using XPath 2.0 which needs to be enabled in the product as shown below before starting the integration service. 
```

>     Add the following to the `PRODUCT-HOME/conf/deployment.toml` file. You can further refer to the [Product Configurations]({{base_path}}/reference/config-catalog-mi/#message-mediation).

Reviewer: ```suggestion
    Add the following to the `<PRODUCT_HOME>/conf/deployment.toml` file. You can further refer to the [Product Configurations]({{base_path}}/reference/config-catalog-mi/#message-mediation).
```

    
      ```
        [mediation]
        synapse.enable_xpath_dom_failover="true"
      ```

    ??? note "Click here for instructions on configuring WSO2 Enterprise Integrator 6"

>         1. Uncomment the `synapse.xpath.dom.failover.enabled=true` property in PRODUCT-HOME/conf/synapse.properties file.

Reviewer: ```suggestion
        1. Uncomment the `synapse.xpath.dom.failover.enabled=true` property in the `<PRODUCT_HOME>/conf/synapse.properties` file.
```

            <enableSSL>true</enableSSL>
        </amazonsqs.init>
    </localEntry>
    ```

3. Navigate to **MI Project Explorer** > **Endpoints** and click on the **+** icon next to **Endpoints**.

4. Select **Address Endpoint**.

    <img src="{{base_path}}/assets/img/integrate/connectors/amazon-sqs/add-endpoint.png" title="Adding an endpoint" width="800" alt="Adding an endpoint"/>

5. You can go to the source view of the XML configuration file of the endpoint and copy the following configuration.
   ```xml
   <?xml version="1.0" encoding="UTF-8"?>
   <endpoint name="SimpleStockQuote" xmlns="http://ws.apache.org/ns/synapse">
       <address     uri="http://localhost:9000/services/SimpleStockQuoteService">
           <suspendOnFailure>
               <initialDuration>-1</initialDuration>
               <progressionFactor>1</progressionFactor>
           </suspendOnFailure>
           <markForSuspension>
               <retriesBeforeSuspension>0</retriesBeforeSuspension>
           </markForSuspension>
       </address>
   </endpoint>
   ```
6. Then create the following sequences: `buildMessage`, `createQueue`, `sendMessage`, and `ReceiveAndForwardMessage`.

> 7. Navigate to **MI Project Explorer** > **Sequences** and click on the **+** sign next to Sequences to open the **Create New Sequence** form.

Reviewer: ```suggestion
7. Navigate to **MI Project Explorer** > **Sequences** and click the **+** sign next to **Sequences** to open the **Create New Sequence** form.
```

            <queueName>{$ctx:queueName}</queueName>
        </amazonsqs.createQueue>
        <property expression="json-eval($.CreateQueueResponse.CreateQueueResult.QueueUrl)" name="queueURL" scope="default" type="STRING"/>
        <log level="custom">
            <property expression="$ctx:queueURL" name="queueURL"/>
        </log>
        <property expression="fn:substring($ctx:queueURL,41,12)" name="queueId" scope="default" type="STRING" xmlns:fn="http://www.w3.org/2005/xpath-functions"/>
        <log level="custom">
            <property expression="$ctx:queueId" name="queueId"/>
        </log>
    </sequence>
  ```

4. Create `sendMessage` sequence as shown below. In this sequence, we send the message that we built in step 5 to the Amazon SQS Queue. 
  ```
  <?xml version="1.0" encoding="UTF-8"?>
    <sequence name="sendMessage" trace="disable" xmlns="http://ws.apache.org/ns/synapse">
        <amazonsqs.sendMessage configKey="AMAZON_SQS_CONNECTION">
            <queueId>{$ctx:queueId}</queueId>
            <queueName>{$ctx:queueName}</queueName>
            <messageBody>{$ctx:target_property}</messageBody>
        </amazonsqs.sendMessage>
    </sequence>
  ```

> 5. Create the ReceiveAndForwardMessage sequence as shown below. In this sequence, we will receive the message from the Amazon SQS queue and forward it to the StockQuote Endpoint. 

Reviewer: ```suggestion
5. Create the `ReceiveAndForwardMessage` sequence as shown below. In this sequence, we will receive the message from the Amazon SQS queue and forward it to the `StockQuote` Endpoint. 
```

    <api context="/sqs" name="SQSAPI" xmlns="http://ws.apache.org/ns/synapse">
        <resource methods="POST" uri-template="/sendToQueue">
            <inSequence>
                <property expression="json-eval($.queueName)" name="queueName" scope="default" type="STRING"/>
                <sequence key="buildMessage"/>
                <sequence key="createQueue"/>
                <sequence key="sendMessage"/>
                <sequence key="ReceiveAndForwardMessage"/>
                <respond/>
            </inSequence>
            <faultSequence/>
        </resource>
    </api>
    ```

Now we can export the artifacts into a Carbon Application (CApp) and deploy it to the server runtime to run it and test it.

## Get the project

You can download the ZIP file and extract the contents to get the project code.

<a href="{{base_path}}/assets/attachments/connectors/amazonSQSExample1.zip">
    <img src="{{base_path}}/assets/img/integrate/connectors/download-zip.png" width="200" alt="Download ZIP">
</a>

!!! tip
    You may need to update the value of the access key and make other such changes before deploying and running this project.

## Deployment

> Follow the steps in [deploy-artifacts]({{base_path}}/develop/deploy-artifacts) guide to build and export it into a CAR file (.car file) using the WSO2 Micro Integrator Visual Studio Code extension.

Reviewer: ```suggestion
Follow the steps in [deploy-artifacts]({{base_path}}/develop/deploy-artifacts) guide to build and export it into a CAR file (`.car` file) using the Micro Integrator for Visual Studio Code extension.
```

## Test

1. Deploy the back-end service `SimpleStockQuoteService`.

    1. Download the ZIP file of the back-end service from [here](https://github.com/wso2-docs/WSO2_EI/blob/master/Back-End-Service/axis2Server.zip).
    2. Extract the downloaded zip file.
    3. Open a terminal and navigate to the `axis2Server/bin/` directory inside the extracted folder.

>     4. Execute the following command to start the axis2server with the SimpleStockQuote back-end service:

Reviewer: ```suggestion
    4. Execute the following command to start the axis2server with the `SimpleStockQuote` back-end service:
```

       === "On MacOS/Linux/CentOS"   
       ```bash
       sh axis2server.sh
       ```
       === "On Windows"              
       ```bash
       axis2server.bat
       ```

3. Invoke the API as shown below using the curl command. Curl Application can be downloaded from [here](https://curl.haxx.se/download.html).

   ```
       curl --location 'http://localhost:8290/sqs/sendToQueue' \
       --header 'Content-Type: application/json' \
       --data '{
          "companyName":"WSO2",
          "queueName":"Queue1"
       }'
    ```

**Expected Response**:

> You should get the following response with the 'sys_id' and keep it saved.

Reviewer: ```suggestion
Next, you will receive a response with `sys_id`. Make sure to save it."
```

```
<ns:getQuoteResponse xmlns:ns="http://services.samples">
    <ns:return xmlns:ax21="http://services.samples/xsd" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:type="ax21:GetQuoteResponse">
        <ax21:change>-2.3656563413900478</ax21:change>
        <ax21:earnings>13.92852425779543</ax21:earnings>
        <ax21:high>161.79828086332458</ax21:high>
        <ax21:last>155.25325499166968</ax21:last>
        <ax21:lastTradeTimestamp>Thu Aug 08 13:28:27 IST 2024</ax21:lastTradeTimestamp>
        <ax21:low>160.3277722524871</ax21:low>
        <ax21:marketCap>5.643236512855213E7</ax21:marketCap>
        <ax21:name>WSO2 Company</ax21:name>
        <ax21:open>-153.68772137674222</ax21:open>
        <ax21:peRatio>-18.75230816155238</ax21:peRatio>
        <ax21:percentageChange>-1.3934902393316715</ax21:percentageChange>
        <ax21:prevClose>169.76483039627422</ax21:prevClose>
        <ax21:symbol>WSO2</ax21:symbol>
        <ax21:volume>19325</ax21:volume>
    </ns:return>
</ns:getQuoteResponse>
```

## What's next

> * To customize this example for your own scenario, see [Amazon SQS Connector Reference Guide]({{base_path}}/reference/connectors/amazonsqs-connector/1.x/amazonsqs-connector-1.x-reference/) documentation for all operation details of the connector.

Reviewer: ```suggestion
* To customize this example for your own scenario, see the [Amazon SQS Connector Reference Guide]({{base_path}}/reference/connectors/amazonsqs-connector/1.x/amazonsqs-connector-1.x-reference/) documentation for detailed information on all connector operations.
```
